contracts/StrategyCurveTwoPool.sol

```diff
@@ -0,0 +1,383 @@
+  uint256 public creditThreshold; // amount of credit in underlying tokens that will automatically trigger a harvest
+
+  /* ========== CONSTRUCTOR ========== */
+
+  constructor(ContractAddress memory _contractsAddress)
+    public
+    BaseStrategy(_contractsAddress.vault)
+  {
+    // Curve contracts
+    gauge = ICurveGauge(_contractsAddress.gauge);
+    gaugeFactory = ICurveGaugeFactory(_contractsAddress.gaugeFactory);
+    curve = ICurvePool(_contractsAddress.pool);
+    crv = IERC20(_contractsAddress.crv);
+
+    // Uniswap V3 router
+    router = ISwapRouter(_contractsAddress.router);
+
+    // Wrapped ETH - Used for multi-hop swap. No contract instance needed
+    wethAddress = _contractsAddress.weth;
+
+    // Arbitrum healthcheck contract address
+    healthCheck = _contractsAddress.healthCheck;
+
+    // Our standard approvals. want = Curve LP token
+        want.approve(_contractsAddress.gauge, type(uint256).max);
+        crv.approve(_contractsAddress.router, type(uint256).max);
+
+    maxReportDelay = 2 days; // 2 days in seconds
+  }
+
+  /* ========== VIEWS ========== */
+
+  function stakedBalance() public view returns (uint256) {
+    return gauge.balanceOf(address(this));
+  }
+
+  function balanceOfWant() public view returns (uint256) {
+    return want.balanceOf(address(this));
+  }
+
+  function estimatedTotalAssets() public view override returns (uint256) {
+    return balanceOfWant().add(stakedBalance());
+  }
+
+  /* ========== MUTATIVE FUNCTIONS ========== */
+  // these should stay the same across different wants.
+
+  function adjustPosition(uint256 _debtOutstanding) internal override {
+    if (emergencyExit) {
+      return;
+    }
+    // Send all of our LP tokens to deposit to the gauge if we have any
+    uint256 _toInvest = balanceOfWant();
+    if (_toInvest > 0) {
+      gauge.deposit(_toInvest);
+    }
+  }
+
+  function liquidatePosition(uint256 _amountNeeded)
+    internal
+    override
+    returns (uint256 _liquidatedAmount, uint256 _loss)
+  {
+    uint256 _wantBal = balanceOfWant();
+    if (_amountNeeded > _wantBal) {
+      // check if we have enough free funds to cover the withdrawal
+      uint256 _stakedBal = stakedBalance();
+      if (_stakedBal > 0) {
+        gauge.withdraw(Math.min(_stakedBal, _amountNeeded.sub(_wantBal)));
+      }
+      uint256 _withdrawnBal = balanceOfWant();
+      _liquidatedAmount = Math.min(_amountNeeded, _withdrawnBal);
+      _loss = _amountNeeded.sub(_liquidatedAmount);
+    } else {
+      // we have enough balance to cover the liquidation available
+      return (_amountNeeded, 0);
+    }
+  }
+
+  function liquidateAllPositions() internal override returns (uint256) {
+    uint256 _stakedBal = stakedBalance();
+    if (_stakedBal > 0) {
+      // don't bother withdrawing zero
+      gauge.withdraw(_stakedBal);
+    }
+    return balanceOfWant();
+  }
+
+    function _claimRewards() internal {
+        gaugeFactory.mint(address(gauge));
```

Comment:
With the new gauge factory, is it possible to see how much a specific user is owed for a gauge before claiming? if so, we could estimate potential rewards, which we're currently not able to with Curve strategies.

Reply:
not on-chain with the current implementation, unfortunately (see [the code here](https://github.com/curvefi/curve-xchain-factory/blob/master/contracts/implementations/ChildGauge.vy#L490)). We do use the function in a few of the tests to estimate rewards before claiming to make sure the gauge is working as expected.